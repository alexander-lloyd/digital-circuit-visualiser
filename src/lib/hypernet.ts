import Assert from "assert";


export type Flat = string;

/**
 * A label on a Hypernet. Either a string e.g. 'f'
 * or can be expanded to another Hypernet.
 * 
 * Similar to a circuit, E.g. an AND gate. We can have
 * represent an AND gate as an integrated circuit (string)
 * or as transistors (Hypernet).
 */
export type Label<T> = Flat | Hyper<T>;

/**
 * A Hypernet can be made of a set of inputs and outputs.
 */
export interface Hyper<T> {
    input: Edge<T>;
    edges: Edge<T>[];
    output: Edge<T>;
}

/**
 * 
 */
export interface Edge<T> {
    sources: [Edge<T>, number][];
    targets: [Edge<T>, number][];
    label: Label<T>;
}

/**
 * Create an id. A wire.
 *
 * @returns Hypernet representing a wire.
 */
export function id<T>(): Hyper<T> {
    const ie: Edge<T> = {
        sources: [],
        targets: [],
        label: ''
    };
    const oe: Edge<T> = {
        sources: [],
        targets: [],
        label: ''
    };

    ie.sources.push([oe, 0]);
    oe.targets.push([ie, 0]);

    return {
        input: ie,
        edges: [],
        output: oe
    };
}

/**
 * Represents a swap in wires.
 *
 * @returns A swap.
 * 
 * A _  _ C
 *    \/
 * B _/\_ D
 */
export function swap<T>(): Hyper<T> {
    const ie: Edge<T> = {
        sources: [],
        targets: [],
        label: ''
    };
    const oe: Edge<T> = {
        sources: [],
        targets: [],
        label: ''
    };
    ie.sources.push([oe, 1], [oe, 0]);
    oe.targets.push([ie, 1], [ie, 0]);

    return {
        input: ie,
        edges: [],
        output: oe
    };
}

/**
 * Compose two hypernets together.
 * 
 * Technically, the labels need to match as well.
 * This has not been implemented.
 * 
 * Careful: this modifies arguments f and g, if you don't want
 * to you need to copy them. 
 *
 * @param f Function f.
 * @param g Function g.
 * @returns Hypernet composing two functions together
 * @throws AssertionError if f.outputs.length == f.inputs.length
 * 
 *   ┌───┐  ┌───┐
 *  ─┤   ├──┤   ├─
 *  ─┤ f ├──┤ g ├─
 *  ─┤   ├──┤   ├─
 *   └───┘  └───┘
 */
export function compose<T>(f: Hyper<T>, g: Hyper<T>): Hyper<T> {
    if (!f || !g) {
        throw Error(`Invalid input ${f},${g} into compose function`);
    }

    // Check output length is equal to input length.
    Assert.equal(f.output.targets.length, g.input.sources.length, 'Input length must be equal to output length');
    for (let i = 0; i < f.output.targets.length; ++i) {
        const [e, k] = f.output.targets[i];
        const [e2, k2] = g.input.sources[i];
        e.sources[k] = [e2, k2];
        e2.targets[k2] = [e, k];
    }

    return {
        input: f.input,
        edges: [...f.edges, ...g.edges],
        output: g.output
    };
}
