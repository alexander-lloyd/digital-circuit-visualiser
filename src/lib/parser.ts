import {Template, template} from './template';
import {SymbolTableStack} from './symboltable';

/**
 * TypeScript does not have a char type.
 * Create an alias for clarity.
 */
type Char = string;

export const EOL: Char = '\n';
export const EOF: Char = '\0';

/**
 * Is value an alphabetic character.
 *
 * @param value The value.
 * @returns true if the character is an alphabetic character.
 */
export function isLetter(value: string): boolean {
    return value.match(/[a-z]/giu) !== null;
}

/**
 * Is value a numeric character?
 *
 * @param value The value.
 * @returns true if the character is a numeric character.
 */
export function isNumeric(value: string): boolean {
    return value.match(/[0-9]/gu) !== null;
}

/**
 * Is the character a whitespace character.
 *
 * @param value The value.
 * @returns true if the value is a whitespace character.
 */
export function isWhitespace(value: string): boolean {
    return value.match(/\s/gu) !== null;
}

/**
 * Source.
 *
 * Represents a source program.
 */
export class Source {
    private readonly source: string[];
    private lineNumber = -1;
    private linePosition = -2;
    private line: string | null = null;

    /**
     * Constructor.
     *
     * @param source The program source.
     */
    public constructor(source: string) {
        this.source = source.split(EOL);
    }

    /**
     * Get the current char.
     *
     * @returns The current char.
     */
    public currentChar(): Char {
        if (this.linePosition === -2) {
            // First time?
            this.readLine();
            return this.nextChar();
        } else if (this.line === null || this.line === undefined) {
            // At end of file?
            return EOF;
        } else if (this.linePosition === -1 || this.linePosition === this.line.length) {
            // At end of line?
            this.readLine();
            return this.nextChar();
        }
        return this.line.charAt(this.linePosition);
    }

    /**
     * Peek at the next character in the source.
     *
     * @returns The next character.
     */
    public peekChar(): Char {
        this.currentChar();

        const nextPos = this.linePosition + 1;

        if (this.line === null) {
            return EOF;
        }

        return nextPos < this.line.length
            ? this.line.charAt(nextPos)
            : EOL;
    }

    /**
     * Get the next character in the source.Increment the position in the string.
     *
     * @returns The next character.
     */
    public nextChar(): Char {
        this.linePosition++;
        return this.currentChar();
    }

    /**
     * Read next line in source.
     */
    private readLine(): void {
        this.lineNumber++;
        this.line = this.source[this.lineNumber];
        this.linePosition = -1;
    }
}

export enum TokenType {
    EoFToken = 'EoFToken',
    SymbolToken = 'SymbolToken',
    WordToken = 'WordToken',
    IdentifierToken = 'IdentifierToken',
    ErrorToken = 'ErrorToken'
}

/**
 * Base class for all Tokens.
 */
export class Token {
    private readonly tokenType: TokenType;

    /**
     * Constructor.
     *
     * @param tokenType The type of token this object is.
     */
    public constructor(tokenType: TokenType) {
        this.tokenType = tokenType;
    }

    /**
     * Get the token type.
     *
     * @returns The token type.
     */
    public getTokenType(): TokenType {
        return this.tokenType;
    }
}

/**
 * An end of file token.
 */
class EoFToken extends Token {
    /**
     * Constructor.
     */
    public constructor() {
        super(TokenType.EoFToken);
    }
}

/**
 * A word token.
 *
 * Represents a literal. E.g. A variable.
 */
class WordToken extends Token {
    private readonly literal: string;

    /**
     * Constructor.
     *
     * @param literal A literal.
     */
    public constructor(literal: string) {
        super(TokenType.WordToken);
        this.literal = literal;
    }

    /**
     * Get the literal.
     *
     * @returns The literal.
     */
    public getLiteral(): string {
        return this.literal;
    }
}

/**
 * A Symbol token.
 *
 * E.g. '+'
 */
class SymbolToken extends Token {
    private readonly symbol: string;

    /**
     * Constructor.
     *
     * @param symbol The symbol.
     */
    public constructor(symbol: string) {
        super(TokenType.SymbolToken);
        this.symbol = symbol;
    }

    /**
     * Get the symbol.
     *
     * @returns The symbol.
     */
    public getSymbol(): string {
        return this.symbol;
    }
}

/**
 * An Identifier.
 *
 * E.g. 'in'
 */
class IdentifierToken extends Token {
    private readonly identifier: string;

    /**
     * Constructor.
     *
     * @param identifier An identifier.
     */
    public constructor(identifier: string) {
        super(TokenType.IdentifierToken);
        this.identifier = identifier;
    }
}

/**
 * Token used when we get an unexpected character.
 */
class ErrorToken extends Token {
    /**
     * Constructor.
     */
    public constructor() {
        super(TokenType.ErrorToken);
    }
}

/**
 * Scanner.
 *
 * Abstract scanner class will be implemented by language-specific classes.
 */
abstract class Scanner {
    protected source: Source;

    /**
     * Constructor
     *
     * @param source The program source.
     */
    public constructor(source: Source) {
        this.source = source;
    }

    /**
     * Extract a token from the Source.
     *
     * @returns The next token.
     */
    public abstract extractToken(): Token;
}

enum ErrorType {
    SYNTAX_ERROR = 'Syntax Error',
    UNEXPECTED_SYMBOL = 'Unexpected Symbol'
}

/**
 * ErrorHandler.
 *
 * Flag warning and errors to the user.
 */
interface ErrorHandler {
    /**
     * Flag an error or warning.
     *
     * @param token The token where the warning should be raised.
     * @param errorType The type of warning or error.
     */
    flag(token: Token, errorType: ErrorType): void;
}

/**
 * Log Warning and Errors to the user using the console.
 */
class ConsoleErrorHandler implements ErrorHandler {
    private template: Template = template`ERROR: ${'errorType'}`;

    /**
     * Flag an error or warning.
     *
     * @param token The token where the warning should be raised.
     * @param errorType The type of warning or error.
     */
    public flag(token: Token, errorType: ErrorType): void {
        console.error(this.template({errorType}));
    }
}

/**
 * Parser.
 *
 * Abstract parser class will be implemented by language-specific classes.
 */
abstract class Parser<T> {
    protected readonly scanner: Scanner;
    protected readonly errorHandler: ErrorHandler;
    protected readonly symbolStack: SymbolTableStack;

    /**
     * Constructor.
     *
     * @param scanner Scanner object.
     * @param errorHandler ErrorHandler object.
     * @param symbolStack Symbol Table Stack.
     */
    public constructor(
        scanner: Scanner,
        errorHandler: ErrorHandler,
        symbolStack: SymbolTableStack
    ) {
        this.scanner = scanner;
        this.errorHandler = errorHandler;
        this.symbolStack = symbolStack;
    }

    /**
     * Parse the program.
     */
    public abstract parse(): T;
}

enum LanguageTokenEnum {
    COMPOSE = '*',
    EQUAL = '=',
}

type LanguageTokenType = TokenType | LanguageTokenEnum;

const SYMBOLS = new Map<string, LanguageTokenEnum>().
    set('=', LanguageTokenEnum.EQUAL).
    set('*', LanguageTokenEnum.COMPOSE);

const RESERVED_WORDS = ['let'];

/**
 * The scanner for this language.
 */
export class LanguageScanner extends Scanner {
    /**
     * Extract a token based on the current character.
     *
     * @returns The next token.
     */
    public extractToken(): Token {
        let token: Token;
        this.skipWhitespace();

        const currentChar: Char = this.source.currentChar();

        if (currentChar === EOF) {
            token = new EoFToken();
        } else if (isLetter(currentChar)) {
            token = this.extractWord();
        } else if (SYMBOLS.has(currentChar)) {
            token = new SymbolToken(currentChar);
        } else {
            token = new ErrorToken();
        }
        this.source.nextChar();

        return token;
    }

    /**
     * Move the source object pointer until the next token.
     */
    private skipWhitespace(): void {
        let currentChar = this.source.currentChar();

        while (isWhitespace(currentChar) && currentChar !== EOF) {
            currentChar = this.source.nextChar();
        }
    }

    /**
     * Extract a word out of the source.
     *
     * @returns A token.
     */
    private extractWord(): Token {
        let currentChar = this.source.currentChar();
        let word = '';

        while (isLetter(currentChar) || isNumeric(currentChar)) {
            word += currentChar;
            currentChar = this.source.nextChar();
        }

        if (RESERVED_WORDS.includes(word)) {
            return new IdentifierToken(word);
        }
        // Not a reserved word so it's just a word.
        return new WordToken(word);
    }
}

/**
 * ASTNodeType.
 */
enum ASTNodeType {
    COMPOSE,
    VARIABLE
}

/**
 * Attributes an AST Node can store.
 */
enum ASTAttribute {
    VARIABLE_NAME,
    VALUE
}

/**
 * AST Node.
 */
interface AST {
    /**
     * Get the type of AST Node.
     *
     * @returns The AST Node type.
     */
    getType(): ASTNodeType;

    /**
     * Add a child to the AST Node.
     *
     * @param node: AST Node.
     */
    addChild(node: AST): void;

    /**
     * Get a attribute from the node.
     *
     * @param key Attribute key.
     * @returns the attribute value.
     */
    getAttribute(key: ASTAttribute): string | undefined;

    /**
     * Set an attribute on the node.
     *
     * @param key The attribute.
     * @param value The attribute value.
     */
    setAttribute(key: ASTAttribute, value: string): void;
}

/**
 * AST Implementation.
 */
class ASTImpl implements AST {
    private readonly nodeType: ASTNodeType;
    private readonly children: AST[] = [];
    private readonly attributes: Map<ASTAttribute, string> = new Map();

    /**
     * Constructor.
     *
     * @param nodeType The AST Node Type.
     */
    public constructor(nodeType: ASTNodeType) {
        this.nodeType = nodeType;
    }

    /**
     * Get the type of AST Node.
     *
     * @returns The AST Node type.
     */
    public getType(): ASTNodeType {
        return this.nodeType;
    }

    /**
     * Add a child to the AST Node.
     *
     * @param node AST Node.
     */
    public addChild(node: AST): void {
        this.children.push(node);
    }

    /**
     * Get a attribute from the node.
     *
     * @param key Attribute key.
     * @returns the attribute value.
     */
    public getAttribute(key: ASTAttribute): string | undefined {
        return this.attributes.get(key);
    }

    /**
     * Set an attribute on the node.
     *
     * @param key The attribute.
     * @param value The attribute value.
     */
    public setAttribute(key: ASTAttribute, value: string): void {
        this.attributes.set(key, value);
    }
}

/**
 * Create an AST Node.
 *
 * @param nodeType AST Node Type.
 * @returns New AST Node.
 */
function createASTNode(nodeType: ASTNodeType): AST {
    return new ASTImpl(nodeType);
}

/**
 * ExpressionParser.
 */
class ExpressionParser extends Parser<AST> {
    private static readonly OPS_MAP: Map<LanguageTokenType, ASTNodeType> = new Map().
        set(LanguageTokenEnum.COMPOSE, ASTNodeType.COMPOSE);

    /**
     * Parse an expression.
     *
     * @returns Expression AST.
     */
    public parse(): AST {
        let token: Token = this.scanner.extractToken();
        let tokenType: TokenType = token.getTokenType();

        if (tokenType !== TokenType.WordToken) {
            this.errorHandler.flag(token, ErrorType.SYNTAX_ERROR);
        }

        // Create root AST
        let rootAST = createASTNode(ASTNodeType.VARIABLE);

        token = this.scanner.extractToken();
        tokenType = token.getTokenType();

        while (tokenType === TokenType.SymbolToken) {
            const symbolToken = token as SymbolToken;
            const symbol = symbolToken.getSymbol();
            const symbolType = SYMBOLS.get(symbol);
            if (symbolType === undefined) {
                this.errorHandler.flag(token, ErrorType.UNEXPECTED_SYMBOL);
                continue;
            }
            const astNodeType = ExpressionParser.OPS_MAP.get(symbolType);
            if (astNodeType === undefined) {
                this.errorHandler.flag(token, ErrorType.UNEXPECTED_SYMBOL);
                continue;
            }
            const ast = createASTNode(astNodeType);
            ast.addChild(rootAST);
            rootAST = ast;
        }

        return rootAST;
    }
}

/**
 * LanguageParser.
 */
export class LanguageParser extends Parser<AST> {
    private readonly expressionParser: ExpressionParser;

    /**
     * Constructor.
     *
     * @param scanner Scanner object.
     * @param errorHandler Error Handler. Called when an error is raised
     * @param symbolStack Symbol Table Stack.
     */
    constructor(scanner: Scanner, errorHandler: ErrorHandler, symbolStack: SymbolTableStack) {
        super(scanner, errorHandler, symbolStack);
        this.expressionParser = new ExpressionParser(scanner, errorHandler, symbolStack);
    }

    /**
     * Parse the program.
     *
     * @returns Expression AST.
     */
    public parse(): AST {
        const ast = this.expressionParser.parse();
        const token = this.scanner.extractToken();

        if (token.getTokenType() !== TokenType.EoFToken) {
            // Error the string isn't finished.
            this.errorHandler.flag(token, ErrorType.SYNTAX_ERROR);
        }

        return ast;
    }
}
