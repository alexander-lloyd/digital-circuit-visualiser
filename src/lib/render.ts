export interface Entity {
}

export interface WireEntity extends Entity {
    x1: number;
    y1: number;
    x2: number;
    y2: number;
}

export interface EntityRenderer<T extends Entity> {
    /**
     * Render an entity to the canvas
     * 
     * @param ctx: Canvas Render Context.
     * @param entity: The entity to render.
     */
    render(ctx: CanvasRenderingContext2D, entity: T): void;
}


export class WireEntityRenderer implements EntityRenderer<WireEntity> {
    render(ctx: CanvasRenderingContext2D, {x1, y1, x2, y2}: WireEntity): void {
        ctx.beginPath();
        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
        ctx.stroke();
    }
}