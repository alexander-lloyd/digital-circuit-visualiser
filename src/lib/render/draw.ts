/* eslint no-magic-numbers: ["warn", {"ignore": [0, 1, 2]}] */
import {
    BoxEntry,
    LineEntry,
    Point
} from './types';

const DEFAULT_CROSS_SIZE = 5;
const DEFAULT_CURVENESS = 50;

/**
 * Used in Debugging.
 *
 * @param point The position to draw the cross.
 * @param size The size of the cross.
 * @returns Line entries to draw the x.
 */
export function drawCross([x, y]: Point, size = DEFAULT_CROSS_SIZE): LineEntry[] {
    const halfSize = size / 2;
    return [
        [[x - halfSize, y - halfSize], [x + halfSize, y + halfSize]],
        [[x - halfSize, y + halfSize], [x + halfSize, y - halfSize]]
    ];
}

/**
 * Work out where on the logic gate image the wires should connect.
 *
 * @param terminatorCount The number of connections.
 * @returns An array of the position along the Y axis in unit space.
 *
 * @example getTerminatorPositions(1) -> 0.5
 * @example getTerminatorPositions(2) -> [0.33, 0.66]
 */
export function getTerminatorPositions(terminatorCount: number): number[] {
    const unitSize = 1.0;
    const array: number[] = new Array(terminatorCount).
        fill(0).
        map((_, index) => unitSize * (index + 1) / (terminatorCount + 1));
    return array;
}

/**
 * Render a line on the Canvas.
 *
 * @param ctx Canvas Render Context.
 * @param line Line to draw.
 */
export function renderLineEntry(
    ctx: CanvasRenderingContext2D,
    [[x1, y1], [x2, y2]]: LineEntry
): void {
    ctx.beginPath();
    ctx.moveTo(x1, y1);
    ctx.lineTo(x2, y2);
    ctx.stroke();
}

/**
 * Render a box on the Canvas.
 *
 * @param ctx Canvas Render Context.
 * @param line Box to draw.
 */
export function renderBoxEntry(
    ctx: CanvasRenderingContext2D,
    [[x, y], [x2, y2]]: BoxEntry
): void {
    const width = Math.abs(x2 - x);
    const height = Math.abs(y2 - y);
    ctx.beginPath();
    ctx.rect(x, y, width, height);
    ctx.stroke();
}

/**
 * Render a cross to the Cavnas. Used for debugging.
 *
 * @param ctx Canvas Render Context.
 * @param point The point to draw the cross.
 * @param size Optional. The size of the cross.
 */
export function renderCross(ctx: CanvasRenderingContext2D, point: Point, size = DEFAULT_CROSS_SIZE): void {
    const lineEntries = drawCross(point, size);
    lineEntries.forEach((line: LineEntry) => renderLineEntry(ctx, line));
}

/**
 * Render a bezier curve.
 *
 * @param ctx Canvas Render Context.
 * @param line Bezier line to draw.
 * @param bezierConstant Bezier Constant
 */
export function renderBezier(
    ctx: CanvasRenderingContext2D,
    [x1, y1]: Point,
    [x2, y2]: Point,
    [cp1x, cp1y]: Point,
    [cp2x, cp2y]: Point
): void {
    ctx.beginPath();
    ctx.moveTo(x1, y1);
    ctx.lineWidth = 2;
    ctx.bezierCurveTo(
        cp1x, cp1y,
        cp2x, cp2y,
        x2, y2
    );
    ctx.stroke();
}

/**
 * Render a curve curve.
 *
 * @param ctx Canvas Render Context.
 * @param line Curve to draw.
 * @param curveness Curveness Constant
 */
export function renderCurve(
    ctx: CanvasRenderingContext2D,
    [[x1, y1], [x2, y2]]: LineEntry,
    curveness = DEFAULT_CURVENESS
): void {
    renderBezier(
        ctx,
        [x1, y1],
        [x2, y2],
        [x1 + curveness, y1],
        [x2 - curveness, y2]
    );
}
