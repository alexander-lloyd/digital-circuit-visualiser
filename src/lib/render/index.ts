import {
    RenderResults
} from './types';
import {
    scaleRenderResult,
    translateRenderResult
} from './transform';
import {
    renderResult,
    Render
} from './render';

export {
    Render,
    RenderResults,
    renderResult,
    scaleRenderResult,
    translateRenderResult
};
