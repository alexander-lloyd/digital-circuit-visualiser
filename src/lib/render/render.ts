/* eslint no-magic-numbers: ["warn", {"ignore": [0.5, 1, 2]}] */
import {
    EntityVisitor,
    FunctionEntity,
    GroupedEntity
} from './entities';
import {
    BoxEntry,
    LabelEntry,
    LineEntry
} from './types';
import {
    scaleLineEntry,
    translateLineEntry
} from './transform';
import {
    renderBoxEntry,
    renderLineEntry,
    renderBezier,
    getTerminatorPositions
} from './draw';
import {RENDER_UNIT_SQUARE} from '../../assets/features';


/**
 * Canvas Context.
 */
export interface CanvasContext {
    canvasWidth: number;
    canvasHeight: number;
}

/**
 * EntityRendererVisitor Context
 */
export interface EntityRendererVisitorContext {
    featureFlags: {[featureId: string]: boolean};
}

export type RenderResults = {
    // Start [x,y], End [x,y]
    lines: LineEntry[];
    // Start [x,y], width, height
    boxes: BoxEntry[];
    // Label functions.
    labels: LabelEntry[];
    // Bezier curves
    beziers: LineEntry[];
};

/**
 * Renderer the entities to the Canvas.
 */
export class EntityRendererVisitor extends EntityVisitor<EntityRendererVisitorContext, RenderResults> {
    /**
     * Renderer a Function to the Canvas.
     *
     * @param entity Function Entity.
     * @param context Visitor Context.
     * @returns All the individual canvas elements to draw.
     */
    visitFunction(entity: FunctionEntity, context: EntityRendererVisitorContext): RenderResults {
        const {x, y, width, height, label, wires, inputs, outputs} = entity;
        const {featureFlags} = context;

        const drawBox = featureFlags[RENDER_UNIT_SQUARE] || false;
        const halfWidth = width / 2;
        const halfHeight = height / 2;

        const functionRenderResult: RenderResults = {
            lines: [...wires],
            boxes: [[[x, y], [x + width, y + height], drawBox]],
            labels: [[label, [x + halfWidth, y + halfHeight], inputs.length, outputs.length]],
            beziers: []
        };

        return functionRenderResult;
    }

    /**
     * Renderer a Group of Entities.
     *
     * @param entity Grouped Entity.
     * @param context Visitor Context.
     * @returns All the individual canvas elements to draw.
     */
    visitGrouped(entity: GroupedEntity, context: EntityRendererVisitorContext): RenderResults {
        const {children} = entity;
        const [left, right] = children;

        const {
            lines: llines,
            boxes: lboxes,
            labels: llabel,
            beziers: lbeziers
        } = left.visit(this, context);
        const {
            lines: rlines,
            boxes: rboxes,
            labels: rlabels,
            beziers: rbeziers
        } = right.visit(this, context);

        return {
            lines: [...llines, ...rlines],
            boxes: [...lboxes, ...rboxes],
            labels: [...llabel, ...rlabels],
            beziers: [...lbeziers, ...rbeziers]
        };
    }
}

/**
 * Scale an entity.
 * A value of 1 wont change coordinates.
 * 2 will double the size.
 *
 * @param renderResults Render Result.
 * @param scaleX X scaling value.
 * @param scaleY Y scaling value.
 * @returns new Render Result.
 */
export function scaleRenderResult(renderResults: RenderResults, scaleX: number, scaleY: number): RenderResults {
    const {lines, boxes, labels, beziers} = renderResults;
    // Lines
    const newLines = lines.map((l: LineEntry) => scaleLineEntry(l, scaleX, scaleY));
    const newBeziers = beziers.map((l: LineEntry) => scaleLineEntry(l, scaleX, scaleY));

    const items: [BoxEntry, LabelEntry][] = boxes.map(([[x, y], [x2, y2], drawBox], i) => {
        const [labelFunc, , inputs, outputs] = labels[i];
        const newX = x * scaleX;
        const newX2 = x2 * scaleX;
        const newY = y * scaleY;
        const newY2 = y2 * scaleY;

        return [
            [[newX, newY], [newX2, newY2], drawBox],
            [labelFunc, [0.5 * (newX2 + newX), 0.5 * (newY2 + newY)], inputs, outputs]
        ];
    });

    const newBoxes = items.map(([b]) => b);
    const newLabels = items.map(([, l]) => l);

    return {
        lines: newLines,
        boxes: newBoxes,
        labels: newLabels,
        beziers: newBeziers
    };
}

/**
 * Translate the position of the entity.
 *
 * @param renderResults Render Result.
 * @param translateX X translation. Positive will move to the right.
 * @param translateY Y translation. Positive will move downwards.
 * @returns New Render Result.
 */
export function transformRenderResult(
    renderResults: RenderResults,
    translateX: number,
    translateY: number
): RenderResults {
    const {lines, boxes, labels, beziers} = renderResults;
    // Lines
    const newLines = lines.map((l: LineEntry) => translateLineEntry(l, translateX, translateY));
    const newBeziers = beziers.map((l: LineEntry) => translateLineEntry(l, translateX, translateY));

    // Boxes
    const newBoxes = boxes.map(([[x, y], [x2, y2], drawBox]): BoxEntry => {
        const newX1 = x + translateX;
        const newX2 = x2 + translateX;
        const newY1 = y + translateY;
        const newY2 = y2 + translateY;
        return [[newX1, newY1], [newX2, newY2], drawBox];
    });

    const newLabels = labels.map(([label, [x, y], inputs, outputs]): LabelEntry => [label, [x + translateX, y + translateY], inputs, outputs]);

    return {
        lines: newLines,
        boxes: newBoxes,
        labels: newLabels,
        beziers: newBeziers
    };
}


/**
 * Render to the Canvas.
 *
 * @param ctx Canvas Rendering Context.
 * @param renderResults Lists of items to render.
 */
export function renderResult(ctx: CanvasRenderingContext2D, renderResults: RenderResults): void {
    const {lines, boxes, labels, beziers} = renderResults;

    // Boxes
    boxes.forEach((box, i) => {
        const [[x, y], [x2, y2], drawBox] = box;
        // This only works because theres a 1 to 1 relation between function boxes and labels.
        const [label, [labelX, labelY], inputs, outputs] = labels[i];
        const width = Math.abs(x2 - x);
        const height = Math.abs(y2 - y);
        // 10% away from edge of box so we can draw lines to connect gates.
        const relatveSeperationFromBox = 0.1;

        if (drawBox) {
            renderBoxEntry(ctx, box);
        }

        const labelWidth = 80;
        const labelHeight = 40;

        label(labelX, labelY, labelWidth, labelHeight, ctx);

        const inputPoints = getTerminatorPositions(inputs);
        const outputPoints = getTerminatorPositions(outputs);

        inputPoints.forEach((inputPos: number) => {
            renderBezier(ctx, [
                [x + (width * relatveSeperationFromBox), y + (height * inputPos)],
                [labelX - (labelWidth / 2), labelY - (labelHeight / 2) + (labelHeight * inputPos)]
            ]);
        });
        outputPoints.forEach((outputPos: number) => {
            renderBezier(ctx, [
                [labelX + (labelWidth / 2), labelY - (labelHeight / 2) + (labelHeight * outputPos)],
                [x2 - (width * relatveSeperationFromBox), y + (height * outputPos)]
            ]);
        });
    });

    // Lines
    lines.forEach((lineEntry: LineEntry): void => renderLineEntry(ctx, lineEntry));
    beziers.forEach((lineEntry: LineEntry): void => renderBezier(ctx, lineEntry));
}
