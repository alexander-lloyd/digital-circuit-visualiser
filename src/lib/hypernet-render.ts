import {Hyper} from './hypernet';

type PositionalData = {
    x: number;
    y: number;
    inputPositions: number[];
    outputPositions: number[];
};

export type HyperPositional = Hyper<PositionalData>;

/**
 * Compute the positions of input and outputs on a function
 * on the unit square.
 *
 * E.g a function with one input should have its input position 0.5 on the left side.
 * We implicitly know its on the input side of the unit square.
 *
 * @param count The number of wires.
 * @returns An array of the position of the wires on the unit square.
 */
export function computePosition(count: number): number[] {
    const unitSize = 1.0;
    const array: number[] = new Array(count).
        fill(0).
        map((_, index) => unitSize * (index + 1) / (count + 1));
    return array;
}

/**
 * Populate the positional data into a hypernet so that it is ready
 * for rendering.
 *
 * Assume the function is in the middle of the square. Then get the
 * number of inputs and output and position them equally along the
 * sides.
 *
 * @param hyper Hypernet.
 * @returns The same hypernet with the populated positional data.
 */
export function populateHyperPositional<T>(hyper: HyperPositional): HyperPositional {
    hyper.data = {
        x: 0.5,
        y: 0.5,
        inputPositions: computePosition(hyper.input.sources.length),
        outputPositions: computePosition(hyper.output.targets.length)

    };

    return hyper;
}
