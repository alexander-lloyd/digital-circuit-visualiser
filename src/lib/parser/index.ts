import {AST} from './ast';
import {parse as parse$} from './parser';
import {
    ASTOptimisingTransformer,
    ASTOptimisingTransformerContext
} from './compile';

/**
 * Parse a string.
 *
 * @param source The source.
 * @returns AST Tree.
 */
export function parse(source: string): AST {
    return parse$(source) as AST;
}

/**
 * Parse and apply optimisations to the AST.
 * @param source Source Code.
 * @returns AST Tree.
 */
export function compile(source: string): AST {
    const ast = parse(source);
    const optimiserContext: ASTOptimisingTransformerContext = {
        identifiers: new Map()
    };
    const optimiser = new ASTOptimisingTransformer();
    const optimised = ast.visit(optimiser, optimiserContext);

    return optimised;
}

export * from './ast';
