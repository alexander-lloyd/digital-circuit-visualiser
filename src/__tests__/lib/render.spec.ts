import { WireEntity, WireEntityRenderer } from '../../lib/render';

describe('WireEntityRenderer', () => {
    it('should match snapshot', () => {
        const canvas: HTMLCanvasElement = document.createElement('canvas');
        const ctx = canvas.getContext('2d') as CanvasRenderingContext2D;

        const wire: WireEntity = {
            x1: 20,
            y1: 20,
            x2: 30,
            y2: 20
          }

        const entity = new WireEntityRenderer();
        entity.render(ctx, wire);

        const events = ctx.__getEvents();
        expect(events).toMatchSnapshot()
        const calls = ctx.__getDrawCalls();
        expect(calls).toMatchSnapshot();
    });
})