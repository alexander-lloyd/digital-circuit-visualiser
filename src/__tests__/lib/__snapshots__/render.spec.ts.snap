// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WireEntityRenderer should match snapshot 1`] = `
Array [
  Object {
    "props": Object {},
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "beginPath",
  },
  Object {
    "props": Object {
      "x": 20,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "moveTo",
  },
  Object {
    "props": Object {
      "x": 30,
      "y": 20,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineTo",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 20,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 30,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;

exports[`WireEntityRenderer should match snapshot 2`] = `
Array [
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 20,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "x": 30,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "lineTo",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
]
`;
